### PR TITLE
Add json format

### PR DESCRIPTION
**Pull request by carakan**
I've been working on this some days ago ... this PR adds support to output json format, json it's a common format to use in external editors or thirdy services, I decided to implement because in my gem "pronto-credo" I'm parsing the flycheck format and I think is not a best approach, keep in consideration this new functionality and if you have some suggestion let me know.

### CODE REVIEW

**Comment by rrrene** [comment]
Hi, sorry for not replying sooner. I am currently swamped at my dayjob, but will try to reply in more detail at the end of the week!

File: lib/credo/cli/output/issues_grouped_by_category.ex [hunk 1]
```diff
@@ -75,11 +78,23 @@ defmodule Credo.CLI.Output.IssuesGroupedByCategory do
       |> Enum.map(&({&1.filename, &1}))
       |> Enum.into(%{})
 
-    categories
-    |> Sorter.ensure(@category_starting_order, @category_ending_order)
-    |> Enum.each(fn(category) ->
-        print_issues(category, issue_map[category], source_file_map, config, term_width)
-      end)
+    case config do
```
Comment (rrrene) [inline lib/credo/cli/output/issues_grouped_by_category.ex]: We shouldn't need a `case` statement here if we utilize pattern matching in method signatures like elsewhere: `%Config{format: "json"} = config`

File: lib/credo/cli/output/issues_grouped_by_category.ex [hunk 2]
```diff
@@ -75,11 +78,23 @@ defmodule Credo.CLI.Output.IssuesGroupedByCategory do
...
+      %Config{format: "json"} ->
+        categories
+        |> Sorter.ensure(@category_starting_order, @category_ending_order)
+        |> Enum.map(fn(category) ->
+            print_issues(category, issue_map[category], source_file_map, config, term_width)
```
Comment (rrrene) [inline lib/credo/cli/output/issues_grouped_by_category.ex]: NOTE: *Never* use a printing function for mapping values.

File: lib/credo/cli/output/issues_grouped_by_category.ex [hunk 3]
```diff
@@ -88,6 +103,12 @@ defmodule Credo.CLI.Output.IssuesGroupedByCategory do
   defp print_issues(_category, nil, _source_file_map, _config, _term_width) do
     nil
   end
+
+  defp print_issues(_category, issues, source_file_map, %Config{format: "json"} = config, term_width) do
+    print_issues(issues, source_file_map, config, term_width)
+    |> List.flatten
```
Comment (rrrene) [inline lib/credo/cli/output/issues_grouped_by_category.ex]: See above: `print_issues` is clearly a method intended for, well, printing issues. Please do not map data for later use here.

File: mix.exs [hunk 4]
```diff
@@ -35,7 +35,8 @@ defmodule Credo.Mixfile do
     [
       {:bunt, "~> 0.1.6"},
       {:inch_ex, "~> 0.5.3", only: [:dev, :test]},
-      {:coverex, "~> 1.4.9", only: :test}
+      {:coverex, "~> 1.4.9", only: :test},
+      {:json,   "~> 0.3.0", optional: true}
```
Comment (rrrene) [inline mix.exs]: Please use Poison as JSON encoder. It's by far the most popular choice. Please also also use a relaxed requirement like `~> 1.5 or ~> 2.0`.

**Comment by rrrene** [comment]
Hi Carlos,

first, thanks for taking the time to put this together. Providing JSON support for other tools is a great idea!

We just need to straighten some things out, before I can merge this. Keep up the good work and please find my comments attached! :+1:

**Comment by carakan** [comment]
Thanks @rrrene for the feedback, I'll review your comments in this week.

**Comment by sstern6** [comment]
@carakan any progress on addressing @rrrene comments?

Also, before you repush could you please rebase against current master?

Thank you!

**Comment by sstern6** [comment]
ping @carakan

**Comment by carakan** [comment]
@sstern6 sorry for the delay, about this, I'll try to rebase and update with last changes in a couple of hours, but I think I would need some help from @rrrene about how to fix/implement a correct approach in some parts of my code.

**Comment by rrrene** [comment]
@carakan Thx for looking into this. There will be a "Credo workshop" at the Hamburg meetup in two weeks. I will bring this issue up and discuss JSON integration into Credo. I will let you know the results of the workshop :D

**Comment by carakan** [comment]
closing this, I haven't time to rebase and continue with this, probably I'll retake on next year 😄

**Pull request closed without merging**